### PR TITLE
twine-pypi 1.8.1 (new formula)

### PR DESCRIPTION
**Pull request by sgerrand**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----
A small 📝 : The test I've added for this formula is not great, but as per [the project's documentation](https://github.com/pypa/twine/blob/master/README.rst), `twine`
> only supports registering projects and uploading distributions

which makes the executable difficult to test beyond `twine --help` or `twine --version`.

### CODE REVIEW

File: Formula/twine-pypi.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    url "https://files.pythonhosted.org/packages/ab/bf/2af6b25f880e2d529a524f98837d33b1048a2a15703fc4806185b54e9672/requests-toolbelt-0.7.1.tar.gz"
+    sha256 "c3843884269d79e492522f3e9f490917e074c1ddbb80111968970e721fe36eaf"
+  end
+
+  def install
+    virtualenv_install_with_resources
+  end
+
+  test do
+    system bin/"twine", "--version"
```
Comment (dunn) [inline Formula/twine-pypi.rb]: Any change of a test that checks more of the functionality?

Comment (sgerrand) [inline Formula/twine-pypi.rb]: 👋 

As per the pull request description:
> A small 📝 : The test I've added for this formula is not great, but as per [the project's documentation](https://github.com/pypa/twine/blob/master/README.rst), `twine`
>> only supports registering projects and uploading distributions
>
> which makes the executable difficult to test beyond `twine --help` or `twine --version`.

I can't easily test `twine register` or `twine upload` as they both actively push Python packages to PyPI repositories and require credentials to do so.

Do you have any suggestions on how I should proceed? I could test `twine register -h` and `twine upload -h`, but that's doesn't feel much better than checking `twine --version`.

Comment (dunn) [inline Formula/twine-pypi.rb]: My mistake, didn't see that comment.  I guess this will do.

File: Formula/twine-pypi.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
+class TwinePypi < Formula
+  include Language::Python::Virtualenv
+
+  desc "Utilities for interacting with PyPI"
+  homepage "https://pypi.python.org"
+  url "https://github.com/pypa/twine/archive/1.8.1.tar.gz"
+  sha256 "918d7ebab52280bac573af9a6c80622202f0cdeab14cb3360f9ffb9848be946d"
+  head "https://github.com/pypa/twine.git"
+
+  depends_on :python
```
Comment (DomT4) [inline Formula/twine-pypi.rb]:
```
depends_on :python => :recommended if MacOS.version <= :snow_leopard
```
Presumably.

Comment (fxcoudert) [inline Formula/twine-pypi.rb]: Actually, we have far more instances of `depends_on :python if MacOS.version <= :snow_leopard` (i.e. without `:recommended`).

Comment (DomT4) [inline Formula/twine-pypi.rb]: Yeah, that seems fine actually. I've always wondered if the `=> :recommended` was to give people a chance to use a custom Python instead of having Homebrew's installed whenever a bottle was poured, but that's not exactly a problem for < 10.6 these days, to say the least.

**Comment by sgerrand** [comment]
👋  @dunn @DomT4 @fxcoudert: So is this good to go now?

**Review by fxcoudert — APPROVED** [review]

**Review by ilovezfs — DISMISSED** [review]
Would prefer we added a real test

**Comment by fxcoudert** [comment]
@ilovezfs see above the discussion (twice) of why a real test cannot be added. I've looked at the package man page and could not figure out a way to add a test that does not actually upload files to the pypi distribution.

**Comment by ilovezfs** [comment]
@fxcoudert I saw. Nonetheless, this needs a real test before it is merged.

**Comment by sgerrand** [comment]
It would be helpful if you could provide a concrete example of what
constitutes a "real test". The twine executable only exposes two "real"
functions, both of which have side effects in PyPI. I'm loathe to invoke
them as part of a test method.

**Comment by ilovezfs** [comment]
This should be fine:
```
diff --git a/Formula/twine-pypi.rb b/Formula/twine-pypi.rb
index 4eed27f..66b7420 100644
--- a/Formula/twine-pypi.rb
+++ b/Formula/twine-pypi.rb
@@ -37,9 +37,12 @@ class TwinePypi < Formula
 
   def install
     virtualenv_install_with_resources
+    pkgshare.install "tests/fixtures/twine-1.5.0-py2.py3-none-any.whl"
   end
 
   test do
-    system bin/"twine", "--version"
+    wheel = "twine-1.5.0-py2.py3-none-any.whl"
+    cmd = "#{bin}/twine upload -uuser -ppass #{pkgshare}/#{wheel} 2>&1"
+    assert_match /Uploading.*#{wheel}.*HTTPError: 403/m, shell_output(cmd, 1)
   end
 end
```

**Comment by sgerrand** [comment]
@ilovezfs: Implemented in a49f533. ♻️

File: Formula/circleci.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
+class Circleci < Formula
```
Comment (fxcoudert) [inline Formula/circleci.rb]: Unrelated file `circleci.rb` needs to be removed.
Also, please squash everything into one commit.

**Review by fxcoudert — CHANGES_REQUESTED** [review]

File: Formula/circleci.rb (hunk 3 above)
Comment (sgerrand) [inline Formula/circleci.rb]: > Unrelated file `circleci.rb` needs to be removed.

Done. 😊

**Review by fxcoudert — APPROVED** [review]

File: Formula/circleci.rb (hunk 3 above)
Comment (sgerrand) [inline Formula/circleci.rb]: > please squash everything into one commit.

Done in 60509a7616. ♻️

File: Formula/twine-pypi.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
+class TwinePypi < Formula
+  include Language::Python::Virtualenv
+
+  desc "Utilities for interacting with PyPI"
+  homepage "https://pypi.python.org"
```
Comment (ilovezfs) [inline Formula/twine-pypi.rb]: What's the reason for using this as the homepage?

Comment (sgerrand) [inline Formula/twine-pypi.rb]: @ilovezfs: None, this has been updated in 197648141e. ♻️

Comment (fxcoudert) [inline Formula/twine-pypi.rb]:
```
  * C: 5: col 3: The homepage should start with http or https (URL is https;//github.com/pypa/twine).
  * The URL https;//github.com/pypa/twine is not reachable
```

Comment (sgerrand) [inline Formula/twine-pypi.rb]: 😊  Fixed in 8d5ef96e5b. ♻️

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
@sgerrand Thanks for the new formula!
